ink-wrapper/src/main.rs

```diff
@@ -253,177 +177,249 @@ fn generate(metadata: &Metadata, types: &HashMap<u32, &Type>) -> rust::Tokens {
         }
 
         impl Instance {
-            $(for constructor in metadata.spec.constructors.iter() {
-                $(define_constructor(&metadata.source.hash, constructor, &types)) $['\n']
+            $(for constructor in metadata.spec().constructors().iter() {
+                $(define_constructor(&code_hash, constructor, metadata)) $['\n']
             })
 
-            $(for message in metadata.spec.messages.iter() {
-                $(define_message(message, &types))
+            $(for message in metadata.spec().messages() {
+                $(define_message(message, metadata))
             })
         }
     }
 }
 
-fn define_type(typ: &Type, types: &HashMap<u32, &Type>) -> rust::Tokens {
-    match &typ.typ.def {
-        TypeDef::Primitive { .. } => rust::Tokens::new(),
-        TypeDef::Variant { variant, .. } => define_variant(typ, &variant, types),
-        TypeDef::Composite { composite, .. } => define_composite(typ, &composite, types),
-        TypeDef::Tuple { .. } => rust::Tokens::new(),
+/// Generates a type definition for a custom type used in the contract.
+fn define_type(typ: &Type<PortableForm>, metadata: &InkProject) -> rust::Tokens {
+    match &typ.type_def() {
+        TypeDef::Variant(variant) => define_variant(typ, variant, metadata),
+        TypeDef::Composite(composite) => define_composite(typ, composite, metadata),
+        _ => quote! {},
```

Comment:
What about [other types](https://docs.rs/scale-info/2.3.1/src/scale_info/ty/mod.rs.html#275-300), like `Array` for example? I tried to use this tool for some `PSP22` contracts IIRC and it failed on this exact case.

Reply:
I just wanted to have this PR be the switch and then add more types in another PR.